tuning/policydrivencontroller/controllermetrics/metrics.py

```diff
@@ -0,0 +1,244 @@
+
+        Returns:
+            dict
+        """
+        if self.__cache == None:
+            logger.warn('EpochLoss cache is NULL!!!!')
+            return None
+        if len(self.__cache) < self.__args['window-size']:
+            logger.info('EpochLoss cache has not grown to size of window yet: %d' % (len(self.__cache)))
+            return None
+        dq = copy.deepcopy(self.__cache)
+        key_prefix = self.__name + "_"
+        exposed_data = {}
+        for i in range(len(dq)):
+            elem = dq.pop()
+            if i == 0:
+                for k, v in elem.items():
+                    key = key_prefix + k + '_epoch_n'
+                    exposed_data[key] =  v
+            else:
+                for k, v in elem.items():
+                    key = key_prefix + k + '_epoch_nm' + str(i)
+                    exposed_data[key] =  v
+        return exposed_data
+
+    def compute(self, training_state, training_args=None, metrics=None):
+        """Computes the controller-metric (epoch-loss over window) and exposes the values of the variables used by the rules.
+
+        Args:
+            training_state: TrainerState object
+            training_args: TrainingArguments object
+            metrics: [optional] metrics data
+
+        Returns:
+            dict
+        """
+        previous_epoch = -1
+        loss_array = None
+        logs_latest_first = list(reversed(training_state.log_history))
+        latest_log_loss = None
+        added = False
+        for i in range(len(logs_latest_first)):
+            log = logs_latest_first[i]
+            try:
+                loss = log['loss']
+                if i == 0:
+                    latest_log_loss = loss
+                epoch = math.ceil(log['epoch'])
+                if previous_epoch == -1:
+                    previous_epoch = epoch
+                    loss_array = []
+                if epoch != previous_epoch and loss_array != None:
+                    if len(loss_array) > 0:
+                        l = np.array(loss_array)
+                        epoch_avg_loss = np.mean(l)
+                        epoch_std_loss = np.std(l, dtype=np.float64)
+                        self.__cache.append({'epoch': epoch, 'avg_loss': epoch_avg_loss, 'std_loss': epoch_std_loss, 'end_loss': latest_log_loss})
+                        added = True
+                        break
+                loss_array.append(loss)
+            except:
+                # Ignoring log lines not containing relevant fields
+                continue
+        if len(loss_array) > 0 and added == False:
+            l = np.array(loss_array)
+            epoch_avg_loss = np.mean(l)
+            epoch_std_loss = np.std(l, dtype=np.float64)
+            self.__cache.append({'epoch': epoch, 'avg_loss': epoch_avg_loss, 'std_loss': epoch_std_loss, 'end_loss': latest_log_loss})
+        return self.__externalize_data()
+        
+            
+class EvalMetricBasedControl(MetricHandler):
```

Comment:
Do we need a separate controller for this? Why not loss be a parameter in the above class?

Reply:
Addressed as discussed.